src/ui/src/mainwindow.cpp

```diff
@@ -1,39 +1,44 @@
-#include <QGridLayout>
-#include <QLabel>
 #include "mainwindow.h"
+
+#include <QFileDialog>
+
 #include "ui_mainwindow.h"
 
 MainWindow::MainWindow(QWidget* parent)
-    : QMainWindow(parent), ui_(new Ui::MainWindow) {
-  ui_->setupUi(this);
-  mdi_area_ = new QMdiArea(this);
-  mdi_area_->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
-  mdi_area_->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
-  setCentralWidget(mdi_area_);
+		: QMainWindow(parent), ui_(new Ui::MainWindow) {
+	ui_->setupUi(this);
+	setWindowTitle(tr("OVERHEAT application"));
+	// renderer widget
+	render_widget_ = new RendererWidget(this);
+	setCentralWidget(render_widget_);
+	render_widget_->showMaximized();
+	// shape widget
+	shape_list_widget_ = new ShapeListWidget(this);
+	addDockWidget(Qt::LeftDockWidgetArea, shape_list_widget_);
+	shape_list_widget_->setWindowTitle("Shapes");
+	// metadata widget
+	metadata_widget_ = new MetadataWidget(this);
+	addDockWidget(Qt::LeftDockWidgetArea, metadata_widget_);
+	metadata_widget_->setWindowTitle("Metadata");
+	// options widget
+	visualization_options_ = new VisualizationOptionsWidget(this);
+	addDockWidget(Qt::LeftDockWidgetArea, visualization_options_);
+	visualization_options_->setWindowTitle("Options");
+	// signals and slots connection
+	connect(ui_->load_file_btn, &QAction::triggered, this,
+	        &MainWindow::OnLoadFileBtnPressed);
 }
 
-MainWindow::~MainWindow() {
-  delete ui_;
-}
+MainWindow::~MainWindow() { delete ui_; }
```

Comment:
move to header plz

Reply:
ui_ definition places in ui_mainwindow.h, that we won't move to header. So type UI::MainWindow is incomplete in header and we can't delete instatnces of it as it cause UB